File: cpp/src/bws/packer/WrapperArray.cpp
#include <bws/packer/WrapperArray.hpp>

#include <array>
#include <utility>
#include <samchon/library/GeneticAlgorithm.hpp>

using namespace std;
using namespace bws::packer;
using namespace samchon::library;
using namespace samchon::protocol;

/* -----------------------------------------------------------
	CONSTRUCTORS
----------------------------------------------------------- */
auto WrapperArray::createChild(shared_ptr<XML>) -> Wrapper*
{
	return new Wrapper();
}

/* -----------------------------------------------------------
	GETTER
----------------------------------------------------------- */
auto WrapperArray::getPrice() const -> double
{
	double price = 0.0;
	for (size_t i = 0; i < size(); i++)
		price += at(i)->getPrice();

	return price;
}
auto WrapperArray::getUtilization() const -> double
{
	if (this->empty() == true)
		return 0.0;

	double numerator = 0.0;
	double denominator = 0.0;

	for (size_t i = 0; i < size(); i++)
	{
		const auto &wrapper = at(i);

		for (size_t j = 0; j < wrapper->size(); j++)
			numerator += wrapper->at(j)->getVolume();

		denominator += wrapper->getContainableVolume();
	}

	return numerator / denominator;
}

/* -----------------------------------------------------------
	EXPORTERS
----------------------------------------------------------- */
auto WrapperArray::TAG() const -> string
{
	return "wrapperArray";
}
auto WrapperArray::CHILD_TAG() const -> string
{
	return "instance";
}

File: ts/src/bws/packer/WrapperArray.ts
namespace bws.packer
{
	export class WrapperArray
		extends samchon.protocol.EntityArray<Wrapper>
	{
		/* -----------------------------------------------------------
			CONSTRUCTORS
		----------------------------------------------------------- */
		/**
		 * Default Constructor.
		 */
		public constructor()
		{
			super();
		}

		/**
		 * @inheritdoc
		 */
		public createChild(xml: samchon.library.XML): Wrapper
		{
			return new Wrapper();
		}

		/* -----------------------------------------------------------
			GETTERS
		----------------------------------------------------------- */
		/**
		 * Get (calculate) price.
		 */
		public getPrice(): number
		{
			let price: number = 0.0;
			for (let i: number = 0; i < this.size(); i++)
				price += this.at(i).getPrice();

			return price;
		}

		/**
		 * Get (calculate) utilization rate.
		 */
		public getUtilization(): number
		{
			if (this.empty() == true)
				return -1.0;

			let numerator: number = 0.0;
			let denominator: number = 0.0;

			for (let i: number = 0; i < this.size(); i++)
			{
				let wrapper: Wrapper = this.at(i);
				
				denominator += wrapper.getContainableVolume();
				for (let j = 0; j < wrapper.size(); j++)
					numerator += wrapper.getContainableVolume();
			}

			return numerator / denominator;
		}

		/* -----------------------------------------------------------
			EXPORTERS
		----------------------------------------------------------- */
		/**
		 * @inheritdoc
		 */
		public TAG(): string
		{
			return "wrapperArray";
		}

		/**
		 * @inheritdoc
		 */
		public CHILD_TAG(): string
		{
			return "instance";
		}
	}
}